pyro/infer/mcmc/util.py

```diff
@@ -0,0 +1,101 @@
+from collections import defaultdict
+
+from pyro.distributions.util import log_sum_exp
+from pyro.infer.util import is_validation_enabled
+from pyro.util import check_site_shape
+
+
+class EnumTraceProbEvaluator(object):
+    """
+    Computes the log probability density of a trace that possibly contains
+    discrete sample sites enumerated in parallel.
+
+    :param model_trace: execution trace from the model.
+    :param bool has_enumerable_sites: whether the trace contains any
+        discrete enumerable sites.
+    :param int max_iarange_nesting: Optional bound on max number of nested
+        :func:`pyro.iarange` contexts.
+    """
+    def __init__(self,
+                 model_trace,
+                 has_enumerable_sites=False,
+                 max_iarange_nesting=float("inf")):
+        self.model_trace = model_trace
+        self.has_enumerable_sites = has_enumerable_sites
+        self.max_iarange_nesting = max_iarange_nesting
+        self.log_probs = defaultdict(list)
+        self._predecessors_cache = {}
+
+    def _get_predecessors_log_factors(self, target_ordinal):
+        """
+        Returns the list of predecessors for `target_ordinal `and
+        their log_prob factors.
+        """
+        if target_ordinal in self._predecessors_cache:
+            return self._predecessors_cache[target_ordinal]
+        log_factors = []
+        predecessors = set()
```

Comment:
Am I correct that `predecessors` is no longer used, so we can remove this data structure?

Reply:
Fixed.